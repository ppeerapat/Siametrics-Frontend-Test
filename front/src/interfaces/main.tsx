export interface IJob extends IObject {
  driver: IDriver;
  date: Date;
  orders: IOrder[];
}

export interface IOrder extends IObject {
  time: string;
  node: string;
  numberOfItem: number;
}

export interface INode extends IObject {
  name: string;
  lat: number;
  lng: number;
}

export interface IDriver extends IObject {
  name: string;
}

interface IObject {
  id: string;
}
