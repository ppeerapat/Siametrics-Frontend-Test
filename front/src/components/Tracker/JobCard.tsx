import { Card } from 'antd';
import React from 'react';
import { IJob } from '../../interfaces/main';

interface JobCardProp {
  job: IJob;
  isActive?: boolean;
}

const JobCard: React.FC<JobCardProp> = ({ job, isActive }) => {
  // const [isSelected, setIsSelected] = React.useState(false);
  return (
    <Card style={{ overflow: 'hidden', backgroundColor: isActive ? '#ffbb96' : 'inherit', cursor: 'pointer' }}>
      <div style={{ display: 'flex' }}>
        Job #
        <p data-testid="job-id" style={{ margin: '0 5px' }}>
          {job.id}
        </p>{' '}
        {job.date}
      </div>{' '}
      <div style={{ display: 'flex' }}>
        Driver: <p data-testid="driver-name">{job.driver.name}</p>
      </div>
    </Card>
  );
};
export default JobCard;
