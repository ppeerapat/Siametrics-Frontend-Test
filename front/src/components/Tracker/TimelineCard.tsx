import { InboxOutlined } from '@ant-design/icons';
import { Timeline } from 'antd';
import React from 'react';
import { IJob, INodes, IOrder } from '../../interfaces/main';

interface Prop {
  job: IJob;
  nodes?: INodes;
  onChange?: (j: IJob, o: IOrder) => void;
}

export const PanelHeader: React.FC<Prop> = ({ job }) => {
  const total = job.orders.reduce((acc, v, i) => {
    return (acc += v.numberOfItem);
  }, 0);
  return (
    <div style={{ display: 'flex', justifyContent: 'space-between', height: '20px' }}>
      <p>
        #{job.id} {job.driver.name}
      </p>
      <div style={{ display: 'flex' }}>
        <p style={{ whiteSpace: 'nowrap' }}>{total}</p>
        <InboxOutlined style={{ fontSize: '20px', color: '#69c0ff', margin: '0 10px' }} />
      </div>
    </div>
  );
};

const TimelineCard: React.FC<Prop> = ({ job, nodes, onChange }) => {
  // const [isSelected, setIsSelected] = React.useState(false);
  const [selectedOrder, setSelectedOrder] = React.useState<IOrder>();
  React.useEffect(() => {
    if (selectedOrder && onChange) {
      onChange(job, selectedOrder);
    }
  }, [selectedOrder]);

  const fromTime = job.orders[0].time.split(':').map((e) => parseInt(e));
  const toTime = job.orders[job.orders.length - 1].time.split(':').map((e) => parseInt(e));
  const totalTime = [toTime[0] - fromTime[0], toTime[1] - fromTime[1]];

  return (
    <>
      <p>Date: {job.date}</p>
      <p>
        From {job.orders[0].time} to {job.orders[job.orders.length - 1].time}
      </p>
      <p>
        Total:{' '}
        {totalTime
          .map((e) =>
            e.toLocaleString('en-US', {
              minimumIntegerDigits: 2,
            })
          )
          .join('hrs ')}
        mins
      </p>
      <Timeline>
        {job.orders.map((e, i) => {
          return (
            <Timeline.Item key={i + e.toString()}>
              <div style={{ display: 'flex', justifyContent: 'space-between', cursor: 'pointer' }} onClick={() => setSelectedOrder({ ...e })}>
                <div>
                  <p>{e.time}</p>
                  <p>Node: {nodes?.[e.node].name}</p>
                </div>
                <div style={{ display: 'flex' }}>
                  <p style={{ whiteSpace: 'nowrap' }}>{e.numberOfItem}</p>
                  <InboxOutlined style={{ fontSize: '20px', color: '#69c0ff', margin: '0 10px' }} />
                </div>
              </div>
            </Timeline.Item>
          );
        })}
      </Timeline>
    </>
  );
};
export default TimelineCard;
