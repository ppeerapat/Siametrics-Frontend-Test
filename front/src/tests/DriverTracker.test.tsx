import { cleanup, render } from '@testing-library/react';
import JobCard from '../components/Tracker/JobCard';
import { IJob, IDriver, IOrder, INodes } from '../interfaces/main';
import ReactDOM from 'react-dom';

import renderer from 'react-test-renderer';
import TimelineCard from '../components/Tracker/TimelineCard';

const driver: IDriver = {
  id: '1',
  name: 'test driver 1',
};
const order: IOrder = {
  id: '1',
  time: '10:00',
  node: '1',
  numberOfItem: 3,
};
const job: IJob = {
  id: '1',
  driver: driver,
  orders: [order],
  date: '2020-10-10',
};

afterEach(cleanup);
describe('Test Render JobCard', () => {
  it('Render JobCard without crashing', () => {
    const div = document.createElement('div');
    ReactDOM.render(<JobCard job={job} />, div);
  });
  it('JobCard Props are correctly rendered', () => {
    const { getByTestId } = render(<JobCard job={job} />);
    expect(getByTestId('job-id')).toHaveTextContent(job.id);
    expect(getByTestId('driver-name')).toHaveTextContent(driver.name);
  });
  it('JobCard matches snapshot', () => {
    const tree = renderer.create(<JobCard job={job} />).toJSON;
    expect(tree).toMatchSnapshot();
  });
});

describe('Test Render TimelineCard', () => {
  const nodes: INodes = { '1': { id: '1', name: 'test Node', lat: 13, lng: 100 } };
  it('Render TimelineCard without crashing', () => {
    const div = document.createElement('div');
    ReactDOM.render(<TimelineCard job={job} />, div);
  });
  it('TimelineCard Props are correctly rendered', () => {
    const { getByTestId } = render(<TimelineCard job={job} />);
    expect(getByTestId('job-date')).toHaveTextContent(job.date);
    expect(getByTestId('job-fromtime')).toHaveTextContent(order.time);
    expect(getByTestId('job-totime')).toHaveTextContent(order.time);
    expect(getByTestId('job-totaltime')).toHaveTextContent('00hrs 00mins');
    // expect(getByTestId('job-orders')).toHaveLength(1);
  });
  it('TimelineCard matches snapshot', () => {
    const tree = renderer.create(<TimelineCard job={job} />).toJSON;
    expect(tree).toMatchSnapshot();
  });
});
