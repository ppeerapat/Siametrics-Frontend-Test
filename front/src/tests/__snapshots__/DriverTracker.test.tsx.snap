// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test Render JobCard JobCard matches snapshot 1`] = `[Function]`;

exports[`Test Render TimelineCard TimelineCard matches snapshot 1`] = `[Function]`;
